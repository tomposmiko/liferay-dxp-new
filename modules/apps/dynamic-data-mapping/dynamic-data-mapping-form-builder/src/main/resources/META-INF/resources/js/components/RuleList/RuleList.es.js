/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import 'clay-button';

import 'clay-label';

import 'clay-dropdown';
import ClayTooltip from 'clay-tooltip';
import {PagesVisitor} from 'dynamic-data-mapping-form-renderer';
import Component from 'metal-component';
import dom from 'metal-dom';
import {EventHandler} from 'metal-events';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import {maxPageIndex, pageOptions} from '../../util/pageSupport.es';
import {getFieldProperty} from '../LayoutProvider/util/fields.es';
import RulesSupport from '../RuleBuilder/RulesSupport.es';
import templates from './RuleList.soy';

/**
 * RuleList.
 * @extends Component
 */

class RuleList extends Component {
	attached() {
		this._eventHandler = new EventHandler();

		this._eventHandler.add(
			dom.on(
				document,
				'mouseup',
				this._handleDocumentMouseUp.bind(this),
				true
			)
		);

		ClayTooltip.init();
	}

	disposeInternal() {
		super.disposeInternal();

		this._eventHandler.removeAllListeners();
	}

	prepareStateForRender(states) {
		const rules = RulesSupport.formatRules(
			this.pages,
			this._setDataProviderNames(states)
		);

		return {
			...states,
			rules: rules.map((rule) => {
				let logicalOperator;
				let invalidRule = false;

				if (rule['logical-operator']) {
					logicalOperator = rule['logical-operator'].toLowerCase();
				}
				else if (rule.logicalOperator) {
					logicalOperator = rule.logicalOperator.toLowerCase();
				}

				invalidRule = RulesSupport.findInvalidRule(this.pages, rule);

				return {
					...rule,
					actions: rule.actions.map((action) => {
						let newAction;

						if (action.action === 'auto-fill') {
							const {inputs, outputs} = action;

							const inputLabel = Object.values(
								inputs
							).map((input) => this._getFieldLabel(input));
							const outputLabel = Object.values(
								outputs
							).map((output) => this._getFieldLabel(output));

							newAction = {
								...action,
								inputLabel,
								outputLabel,
							};
						}
						else if (action.action == 'jump-to-page') {
							newAction = {
								...action,
								label: this._getJumpToPageLabel(rule, action),
							};
						}
						else {
							newAction = {
								...action,
								label: this._getFieldLabel(action.target),
							};
						}

						return newAction;
					}),
					conditions: rule.conditions.map((condition) => {
						return {
							...condition,
							operands: condition.operands.map(
								(operand, index) => {
									const label = this._getOperandLabel(
										condition.operands,
										index
									);

									return {
										...operand,
										label,
										value: label,
									};
								}
							),
						};
					}),
					invalidRule,
					logicalOperator,
					rulesCardOptions: this._getRulesCardOptions(rule),
				};
			}),
		};
	}

	_getDataProviderName(id) {
		const {dataProvider} = this;

		return dataProvider.find((data) => data.uuid == id).label;
	}

	_getFieldLabel(fieldName) {
		const pages = this.pages;

		return getFieldProperty(pages, fieldName, 'label') || fieldName;
	}

	_getJumpToPageLabel(rule, action) {
		const {pages} = this;
		let pageLabel = '';

		const fieldTarget = (parseInt(action.target, 10) + 1).toString();
		const maxPageIndexRes = maxPageIndex(rule.conditions, pages);
		const pageOptionsList = pageOptions(pages, maxPageIndexRes);
		const selectedPage = pageOptionsList.find((option) => {
			return option.value == fieldTarget;
		});

		if (selectedPage) {
			pageLabel = selectedPage.label;
		}

		return pageLabel;
	}

	_getOperandLabel(operands, index) {
		let label = '';
		const operand = operands[index];

		if (operand.type === 'field') {
			label = this._getFieldLabel(operand.value);
		}
		else if (operand.type === 'user') {
			label = Liferay.Language.get('user');
		}
		else if (operand.type !== 'field') {
			const fieldType = RulesSupport.getFieldType(
				operands[0].value,
				this.pages
			);

			if (
				fieldType === 'checkbox_multiple' ||
				fieldType === 'radio' ||
				fieldType === 'select'
			) {
				label = this._getOptionLabel(operands[0].value, operand.value);
			}
			else if (operand.type === 'json') {
				label = '';

				const operandValueJSON = JSON.parse(operand.value);

				for (const key in operandValueJSON) {
					const keyLabel = this._getPropertyLabel(
						operands[0].value,
						'rows',
						key
					);

					const valueLabel = this._getPropertyLabel(
						operands[0].value,
						'columns',
						operandValueJSON[key]
					);

					label += keyLabel + ':' + valueLabel + ', ';
				}

				const lastCommaPosition = label.lastIndexOf(', ');

				if (lastCommaPosition != -1) {
					label = label.substr(0, lastCommaPosition);
				}
			}
			else {
				label = operand.value;
			}
		}
		else {
			label = operand.value;
		}

		return label;
	}

	_getOptionLabel(fieldName, optionValue) {
		return this._getPropertyLabel(fieldName, 'options', optionValue);
	}

	_getPropertyLabel(fieldName, propertyName, propertyValue) {
		const pages = this.pages;

		let fieldLabel = null;

		if (pages && propertyValue) {
			const visitor = new PagesVisitor(pages);

			visitor.findField((field) => {
				let found = false;

				if (field.fieldName === fieldName && field.options) {
					field[propertyName].some((property) => {
						if (property.value == propertyValue) {
							fieldLabel = property.label;

							found = true;
						}

						return found;
					});
				}

				return found;
			});
		}

		return fieldLabel ? fieldLabel : propertyValue;
	}

	_getRulesCardOptions(rule) {
		const hasNestedCondition = this._hasNestedCondition(rule);

		const rulesCardOptions = [
			{
				disabled: hasNestedCondition,
				label: Liferay.Language.get('edit'),
				settingsItem: 'edit',
			},
			{
				confirm: hasNestedCondition,
				label: Liferay.Language.get('delete'),
				settingsItem: 'delete',
			},
		];

		return rulesCardOptions;
	}

	_handleDocumentMouseUp({target}) {
		const dropdownSettings = dom.closest(target, '.ddm-rule-list-settings');

		if (dropdownSettings) {
			return;
		}

		this.setState({
			dropdownExpandedIndex: -1,
		});
	}

	_handleDropdownClicked(event) {
		event.preventDefault();

		const {dropdownExpandedIndex} = this;
		const ruleNode = dom.closest(event.delegateTarget, '.component-action');

		let ruleIndex = parseInt(ruleNode.dataset.ruleIndex, 10);

		if (ruleIndex === dropdownExpandedIndex) {
			ruleIndex = -1;
		}

		this.setState({
			dropdownExpandedIndex: ruleIndex,
		});
	}

	_handleRuleCardClicked({data, target}) {
		const cardElement = dom.closest(target.element, '[data-card-id]');
		const cardId = parseInt(cardElement.getAttribute('data-card-id'), 10);

		if (data.item.settingsItem == 'edit') {
			this.emit('ruleEdited', {
				ruleId: cardId,
			});
		}
		else if (data.item.settingsItem == 'delete') {
			if (
				!data.item.confirm ||
				confirm(
					Liferay.Language.get(
						'you-cannot-create-rules-with-nested-functions.-are-you-sure-you-want-to-delete-this-rule'
					)
				)
			) {
				this.emit('ruleDeleted', {
					ruleId: cardId,
				});
			}
		}
	}

	_hasNestedCondition(rule) {
		return (
			rule.conditions.find((condition) =>
				condition.operands.find((operand) =>
					operand.value.match(/[aA-zZ]+[(].*[,]+.*[)]/)
				)
			) !== undefined
		);
	}

	_setDataProviderNames(states) {
		const {rules} = states;

		if (this.dataProvider) {
			for (let rule = 0; rule < rules.length; rule++) {
				const actions = rules[rule].actions;

				actions.forEach((action) => {
					if (action.action === 'auto-fill') {
						const dataProviderName = this._getDataProviderName(
							action.ddmDataProviderInstanceUUID
						);

						action.dataProviderName = dataProviderName;
					}
				});
			}
		}

		return rules;
	}
}

RuleList.STATE = {
	dataProvider: Config.arrayOf(
		Config.shapeOf({
			id: Config.string(),
			name: Config.string(),
			uuid: Config.string(),
		})
	),

	dropdownExpandedIndex: Config.number().internal(),

	pages: Config.array().required(),

	roles: Config.arrayOf(
		Config.shapeOf({
			id: Config.string(),
			name: Config.string(),
		})
	).value([]),

	/**
	 * @default 0
	 * @instance
	 * @memberof RuleList
	 * @type {?array}
	 */

	rules: Config.arrayOf(
		Config.shapeOf({
			actions: Config.arrayOf(
				Config.shapeOf({
					action: Config.string(),
					ddmDataProviderInstanceUUID: Config.string(),
					expression: Config.string(),
					inputs: Config.object(),
					label: Config.string(),
					outputs: Config.object(),
					target: Config.string(),
				})
			),
			conditions: Config.arrayOf(
				Config.shapeOf({
					operands: Config.arrayOf(
						Config.shapeOf({
							label: Config.string(),
							repeatable: Config.bool(),
							type: Config.string(),
							value: Config.string(),
						})
					),
					operator: Config.string(),
				})
			),
			['logical-operator']: Config.string(),
		})
	).value([]),

	/**
	 * @default undefined
	 * @instance
	 * @memberof RuleList
	 * @type {!string}
	 */

	spritemap: Config.string().required(),

	/**
	 * @default strings
	 * @instance
	 * @memberof RuleList
	 * @type {!object}
	 */

	strings: Config.object().value({
		'belongs-to': Liferay.Language.get('belongs-to'),
		'calculate-field': Liferay.Language.get('calculate-field-x-as-x'),
		contains: Liferay.Language.get('contains'),
		'due-to-missing-fields': Liferay.Language.get('due-to-missing-fields'),
		'equals-to': Liferay.Language.get('is-equal-to'),
		'greater-than': Liferay.Language.get('is-greater-than'),
		'greater-than-equals': Liferay.Language.get(
			'is-greater-than-or-equal-to'
		),
		'is-empty': Liferay.Language.get('is-empty'),
		'less-than': Liferay.Language.get('is-less-than'),
		'less-than-equals': Liferay.Language.get('is-less-than-or-equal-to'),
		'not-contains': Liferay.Language.get('does-not-contain'),
		'not-equals-to': Liferay.Language.get('is-not-equal-to'),
		'not-is-empty': Liferay.Language.get('is-not-empty'),
	}),
};

Soy.register(RuleList, templates);

export default RuleList;
