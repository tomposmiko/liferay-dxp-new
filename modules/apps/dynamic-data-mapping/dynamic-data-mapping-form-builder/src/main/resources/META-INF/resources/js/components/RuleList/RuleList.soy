{namespace LayoutRule}

/**
 * Prints the DDM form rule builder cotainer.
 */
{template .render}
	{@param? dropdownExpandedIndex: number}
	{@param rules: list<[
			conditions: list<[
				operator: string,
				operands: list<[
					type: string,
					label: string,
					value: string
				]>
			]>,
			actions: list<[
				action: string,
				dataProviderName: ?,
				expression: ?,
				inputLabel: ?,
				target: string,
				label: string,
				outputLabel: ?
			]>,
			logicalOperator: string
		]>
	}
	{@param? rulesCardOptions: ?}
	{@param spritemap: string}
	{@param strings: map<string, string>}
	{@param _handleDropdownClicked: any}
	{@param _handleRuleCardClicked: any}

	<div class="form-builder-rule-builder-container">
		<h1 class="form-builder-section-title text-default">{msg desc=""}rule-builder{/msg}</h1>

		<div class="liferay-ddm-form-rule-rules-list-container">
			{call .rule_list}
				{param _handleDropdownClicked: $_handleDropdownClicked /}
				{param _handleRuleCardClicked: $_handleRuleCardClicked /}
				{param dropdownExpandedIndex: $dropdownExpandedIndex /}
				{param rules: $rules /}
				{param rulesCardOptions: $rulesCardOptions /}
				{param spritemap: $spritemap /}
				{param strings: $strings /}
			{/call}
		</div>
	</div>
{/template}

/**
 * Prints the DDM form card rule.
 */
{template .rule_list}
	{@param rules: list<[
			conditions: list<[
				operator: string,
				operands: list<[
					type: string,
					label: string,
					value: ?
				]>
			]>,
			actions: list<[
				action: string,
				dataProviderName: ?,
				expression: ?,
				inputLabel: ?,
				target: string,
				label: string,
				outputLabel: ?
			]>,
			logicalOperator: string
		]>
	}
	{@param spritemap: string}
	{@param strings: map<string, string>}
	{@param _handleDropdownClicked: any}
	{@param _handleRuleCardClicked: any}
	{@param? dropdownExpandedIndex: number}
	{@param? rulesCardOptions: ?}

	<div class="ddm-rule-list-container form-builder-rule-list">
		{if length($rules) > 0}
			<ul class="ddm-form-body-content form-builder-rule-builder-rules-list tabular-list-group">
				{foreach $rule in $rules}
					<li class="list-group-item">
						<div class="clamp-horizontal list-group-item-content">
							<p class="form-builder-rule-builder-rule-description text-default">
								<b>{msg desc=""}if{/msg} </b>

								{foreach $condition in $rule.conditions}
									{call .condition}
										{param operandType: $condition.operands[0].type /}
										{param operandValue: $condition.operands[0].value /}
									{/call}

									<b class="text-lowercase"><em> {$strings[$condition.operator]} </em></b>

									{if $condition.operands[1]}
										{call .condition}
											{param operandLabel: $condition.operands[1].label /}
											{param operandType: $condition.operands[1].type /}
											{param operandValue: $condition.operands[1].value /}
										{/call}
									{/if}

									{if not isLast($condition)}
										<br />

										<b>
											{' '}
											{if $rule.logicalOperator == 'and'}
												{msg desc=""}and{/msg}
											{else}
												{msg desc=""}or{/msg}
											{/if}
											{' '}
										</b>
									{/if}
								{/foreach}

								<br />

								{foreach $action in $rule.actions}
									{call .action}
										{param action: $action.action /}
										{param dataProviderName: $action.dataProviderName /}
										{param inputLabel: $action.inputLabel /}
										{param expression: $action.expression /}
										{param label: $action.label /}
										{param logicalOperator: $rule.logicalOperator /}
										{param outputLabel: $action.outputLabel /}
									{/call}

									{if not isLast($action)}
									, <br />
									<b> {msg desc=""}and{/msg} </b>
									{/if}
								{/foreach}
							</p>
						</div>

						<div class="list-group-item-field">
							<div class="card-col-field">
								<div class="dropdown dropdown-action">
									{let $ruleIndex: index($rule) /}
									{let $isExpanded: $dropdownExpandedIndex == $ruleIndex /}

									<div class="ddm-rule-list-settings" data-card-id="{$ruleIndex}">
										{call ClayActionsDropdown.render}
											{param items: $rulesCardOptions /}
											{param expanded: $isExpanded /}
											{param events: [
												'itemClicked': $_handleRuleCardClicked
											] /}
											{param spritemap: $spritemap /}
										{/call}

										<button class="component-action cursor-pointer dropdown-toggle" data-onclick="{$_handleDropdownClicked}" data-rule-index="{$ruleIndex}">
											{call ClayIcon.render}
												{param spritemap: $spritemap /}
												{param symbol: 'ellipsis-v' /}
											{/call}
										</button>
									</div>
								</div>
							</div>
						</div>
					</li>
				{/foreach}
			</ul>
		{else}
			{call .empty_list}
				{param message kind="text"}
					{msg desc=""}there-are-no-rules-yet-click-on-plus-icon-below-to-add-the-first{/msg}
				{/param}
			{/call}
		{/if}
	</div>
{/template}

/**
 * Prints the DDM form card rule.
 */
{template .empty_list}
	{@param? message: string}

	<div class="main-content-body">
		<div class="sheet taglib-empty-result-message">
			<div class="taglib-empty-result-message-header"></div>

			{if $message}
				<div class="sheet-text text-center text-muted">
					<p class="text-default">{$message}</p>
				</div>
			{/if}
		</div>
	</div>
{/template}

/**
 * Prints the DDM form card rule.
 */
{template .label}
	{@param? content: string}

	<span class="label label-lg label-secondary" data-original-title="{$content}" title="{$content}">
		<span class="text-truncate-inline">
			<span class="text-truncate">{$content}</span>
		</span>
	</span>
{/template}

/**
 * Prints Rules Conditions.
 */
{template .condition}
	{@param operandType: string}
	{@param? operandLabel: string}
	{@param? operandValue: ?}

	<div class="ddm-condition-container">
		{if $operandType == 'double' or $operandType == 'integer' or $operandType == 'string' or $operandType == 'text' or $operandType == 'numeric'}
			<span>{msg desc=""}value{/msg} </span>
		{elseif $operandType == 'user'}
			<span>{msg desc=""}user{/msg} </span>
		{elseif $operandType == 'field'}
			<span>{msg desc=""}field{/msg} </span>
		{elseif $operandType == 'list'}
			<span>{msg desc=""}list{/msg} </span>
		{/if}

		{if $operandLabel}
			{call .label}
				{param content: $operandLabel /}
			{/call}
		{else}
			{call .label}
				{param content: $operandValue /}
			{/call}
		{/if}
	</div>
{/template}

/**
 * Prints the show action.
 */
{template .action}
	{@param action: string}
	{@param dataProviderName: ?}
	{@param expression: ?}
	{@param label: ?}
	{@param outputLabel: ?}

	{if $action == 'auto-fill'}
		{let $dataProviderOutputFields kind="html"}
			{foreach $out in $outputLabel}
				{call .label}
					{param content: $out /}
				{/call}
				{if not isLast($out)}, {/if}
			{/foreach}
		{/let}

		<span><b>{msg desc=""}autofill-{$dataProviderOutputFields}-from-data-provider-{$dataProviderName}{/msg}</b></span>
	{elseif $action == 'calculate'}
		{let $expressionLabel kind="html"}
			{call .label}
				{param content: $expression /}
			{/call}
		{/let}
		{let $targetLabel kind="html"}
			{call .label}
				{param content: $label /}
			{/call}
		{/let}

		<span><b>{msg desc=""}calculate-field-{$expressionLabel}-as-{$targetLabel}{/msg}</b></span>
	{elseif $action == 'enable'}
		<span><b>{msg desc=""}enable{/msg} </b></span>

		{call .label}
			{param content: $label /}
		{/call}
	{elseif $action == 'jump-to-page'}
		<span><b>{msg desc=""}jump-to-page{/msg} </b></span>

		{call .label}
			{param content: $label /}
		{/call}
	{elseif $action == 'require'}
		<span><b>{msg desc=""}require{/msg} </b></span>

		{call .label}
			{param content: $label /}
		{/call}
	{elseif $action == 'show'}
		<span><b>{msg desc=""}show{/msg} </b></span>

		{call .label}
			{param content: $label /}
		{/call}
	{/if}
{/template}