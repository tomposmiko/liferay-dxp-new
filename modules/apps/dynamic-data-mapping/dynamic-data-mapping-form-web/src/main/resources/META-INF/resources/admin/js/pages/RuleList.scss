.form-rule-list {
	h1 {
		font-size: 32px;
		font-weight: 300;
		margin: 0;
		padding-top: 40px;
	}

	.form-rule-list-col {
		display: flex;
		width: max-content;

		.form-rule-list-invalid-rule {
			padding-top: 3px;
		}
	}

	.list-group {
		font-size: 16px;
	}

	.inline-item {
		max-width: 100%;
	}

	.inline-item-after {
		margin-left: 0.3rem;
	}

	.inline-item-before {
		margin-right: 0.3rem;
	}

	.text-truncate-inline {
		.text-truncate {
			white-space: normal;
			word-wrap: break-word;
		}
	}
}
