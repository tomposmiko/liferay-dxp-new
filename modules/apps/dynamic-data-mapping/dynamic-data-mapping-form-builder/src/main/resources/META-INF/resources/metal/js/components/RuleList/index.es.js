import RuleList from './RuleList.es';

export default RuleList;