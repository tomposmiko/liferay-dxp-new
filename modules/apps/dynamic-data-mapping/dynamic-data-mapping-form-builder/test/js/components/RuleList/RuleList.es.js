/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import RuleList from '../../../../src/main/resources/META-INF/resources/js/components/RuleList/RuleList.es';

let component;

const spritemap = 'icons.svg';

const pages = [
	{
		rows: [
			{
				columns: [
					{
						fields: [
							{
								fieldName: 'text1',
								label: 'label text 1',
								type: 'text',
							},
							{
								fieldName: 'text2',
								label: 'label text 2',
								type: 'text',
							},
						],
					},
				],
			},
		],
	},
];

const brokenRuleConfig = {
	pages,
	rules: [
		{
			actions: [
				{
					action: 'require',
					label: 'label text 1',
					target: 'text1',
				},
			],
			conditions: [
				{
					operands: [
						{
							type: 'field',
							value: 'text1',
						},
						{
							type: '',
							value: '',
						},
					],
					operator: 'contains',
				},
				{
					operands: [
						{
							type: 'field',
							value: 'text1',
						},
						{
							type: 'field',
							value: 'text2',
						},
					],
					operator: 'equals-to',
				},
			],
			['logical-operator']: 'OR',
		},
	],
	spritemap,
};

const configDefault = {
	pages,
	rules: [
		{
			actions: [
				{
					action: 'require',
					label: 'label text 1',
					target: 'text1',
				},
			],
			conditions: [
				{
					operands: [
						{
							type: 'field',
							value: 'text1',
						},
						{
							type: 'value',
							value: 'value 2',
						},
					],
					operator: 'contains',
				},
				{
					operands: [
						{
							type: 'field',
							value: 'text1',
						},
						{
							type: 'field',
							value: 'text2',
						},
					],
					operator: 'equals-to',
				},
			],
			['logical-operator']: 'OR',
		},
		{
			actions: [
				{
					action: 'show',
					label: 'label text 2',
					target: 'text2',
				},
			],
			conditions: [
				{
					operands: [
						{
							type: 'field',
							value: 'text1',
						},
						{
							type: 'value',
							value: 'value 3',
						},
					],
					operator: 'not-equals-to',
				},
			],
			['logical-operator']: 'AND',
		},
	],
	spritemap,
};

describe('RuleList', () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('returns the field label for each action', () => {
		component = new RuleList(configDefault);

		const contextLabel =
			component.pages[0].rows[0].columns[0].fields[0].label;

		const actionLabel = component.rules[0].actions[0].label;

		jest.runAllTimers();

		expect(actionLabel).toEqual(contextLabel);
	});

	it('shows message when rule list is empty', () => {
		component = new RuleList({
			pages,
			rules: [],
			spritemap,
			strings: {
				emptyListText:
					'there-are-no-rules-yet-click-on-plus-icon-below-to-add-the-first',
			},
		});

		expect(component).toMatchSnapshot();
	});

	it('shows rule list', () => {
		component = new RuleList(configDefault);

		expect(component).toMatchSnapshot();
	});

	it('shows the label broken rule when a rule is incomplete', () => {
		component = new RuleList(brokenRuleConfig);

		expect(component).toMatchSnapshot();
	});

	it('shows rule with json type operand', () => {
		const rule = {
			actions: [
				{
					action: 'require',
					label: 'label text 1',
					target: 'text1',
				},
			],
			conditions: [
				{
					operands: [
						{
							type: 'field',
							value: 'grid1',
						},
						{
							type: 'json',
							value: '{"value1" : "value2"}',
						},
					],
					operator: 'equals-to',
				},
			],
			['logical-operator']: 'OR',
		};

		const config = {
			pages: [
				{
					rows: [
						{
							columns: [
								{
									fields: [
										{
											fieldName: 'grid1',
											label: 'label grid 1',
										},
										{
											fieldName: 'text1',
											label: 'label text 1',
										},
									],
								},
							],
						},
					],
				},
			],
			rules: [rule],
			spritemap,
		};

		component = new RuleList(config);

		expect(component._getOperandLabel(rule.conditions[0].operands, 1)).toBe(
			'value1:value2'
		);
	});
});
