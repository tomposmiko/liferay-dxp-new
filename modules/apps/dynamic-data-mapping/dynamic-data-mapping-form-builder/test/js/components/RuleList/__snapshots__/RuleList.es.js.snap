// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuleList shows message when rule list is empty 1`] = `
<div class="form-builder-rule-builder-container">
  <h1 class="form-builder-section-title text-default">rule-builder</h1>
  <div class="liferay-ddm-form-rule-rules-list-container">
    <div class="ddm-rule-list-container form-builder-rule-list">
      <div class="main-content-body">
        <div class="sheet taglib-empty-result-message">
          <div class="taglib-empty-result-message-header"></div>
          <div class="sheet-text text-center text-muted">
            <p class="text-default">there-are-no-rules-yet-click-on-plus-icon-below-to-add-the-first</p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`RuleList shows rule list 1`] = `
<div class="form-builder-rule-builder-container">
  <h1 class="form-builder-section-title text-default">rule-builder</h1>
  <div class="liferay-ddm-form-rule-rules-list-container">
    <div class="ddm-rule-list-container form-builder-rule-list">
      <ul class="ddm-form-body-content form-builder-rule-builder-rules-list tabular-list-group">
        <li class="list-group-item">
          <div class="clamp-horizontal list-group-item-content">
            <p class="form-builder-rule-builder-rule-description text-default"><b>if </b>
              <div class="ddm-condition-container"><span>field </span><span class="label label-lg label-secondary" data-original-title="label text 1" title="label text 1"><span class="text-truncate-inline"><span class="text-truncate">label text 1</span></span></span></div><b class="text-lowercase"><em> contains </em></b>
              <div class="ddm-condition-container"><span class="label label-lg label-secondary" data-original-title="value 2" title="value 2"><span class="text-truncate-inline"><span class="text-truncate">value 2</span></span></span></div><br><b> or </b>
              <div class="ddm-condition-container"><span>field </span><span class="label label-lg label-secondary" data-original-title="label text 1" title="label text 1"><span class="text-truncate-inline"><span class="text-truncate">label text 1</span></span></span></div><b class="text-lowercase"><em> is-equal-to </em></b>
              <div class="ddm-condition-container"><span>field </span><span class="label label-lg label-secondary" data-original-title="label text 2" title="label text 2"><span class="text-truncate-inline"><span class="text-truncate">label text 2</span></span></span></div><br><span><b>require </b></span><span class="label label-lg label-secondary" data-original-title="label text 1" title="label text 1"><span class="text-truncate-inline"><span class="text-truncate">label text 1</span></span></span>
            </p>
          </div>
          <div class="list-group-item-field">
            <div class="card-col-field">
              <div class="dropdown dropdown-action">
                <div class="ddm-rule-list-settings" data-card-id="0">
                  <div class="dropdown dropdown-action"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" title="actions" type="button"><svg class="lexicon-icon lexicon-icon-ellipsis-v" focusable="false" role="presentation">
                        <use xlink:href="icons.svg#ellipsis-v"></use>
                      </svg></button></div><button class="component-action cursor-pointer dropdown-toggle" data-rule-index="0"><svg class="lexicon-icon lexicon-icon-ellipsis-v" focusable="false" role="presentation">
                      <use xlink:href="icons.svg#ellipsis-v"></use>
                    </svg></button>
                </div>
              </div>
            </div>
          </div>
        </li>
        <li class="list-group-item">
          <div class="clamp-horizontal list-group-item-content">
            <p class="form-builder-rule-builder-rule-description text-default"><b>if </b>
              <div class="ddm-condition-container"><span>field </span><span class="label label-lg label-secondary" data-original-title="label text 1" title="label text 1"><span class="text-truncate-inline"><span class="text-truncate">label text 1</span></span></span></div><b class="text-lowercase"><em> is-not-equal-to </em></b>
              <div class="ddm-condition-container"><span class="label label-lg label-secondary" data-original-title="value 3" title="value 3"><span class="text-truncate-inline"><span class="text-truncate">value 3</span></span></span></div><br><span><b>show </b></span><span class="label label-lg label-secondary" data-original-title="label text 2" title="label text 2"><span class="text-truncate-inline"><span class="text-truncate">label text 2</span></span></span>
            </p>
          </div>
          <div class="list-group-item-field">
            <div class="card-col-field">
              <div class="dropdown dropdown-action">
                <div class="ddm-rule-list-settings" data-card-id="1">
                  <div class="dropdown dropdown-action"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" title="actions" type="button"><svg class="lexicon-icon lexicon-icon-ellipsis-v" focusable="false" role="presentation">
                        <use xlink:href="icons.svg#ellipsis-v"></use>
                      </svg></button></div><button class="component-action cursor-pointer dropdown-toggle" data-rule-index="1"><svg class="lexicon-icon lexicon-icon-ellipsis-v" focusable="false" role="presentation">
                      <use xlink:href="icons.svg#ellipsis-v"></use>
                    </svg></button>
                </div>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="edit" type="button">edit</button></li>
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="delete" type="button">delete</button></li>
  </ul>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="edit" type="button">edit</button></li>
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="delete" type="button">delete</button></li>
  </ul>
</div>
`;

exports[`RuleList shows the label broken rule when a rule is incomplete 1`] = `
<div class="form-builder-rule-builder-container">
  <h1 class="form-builder-section-title text-default">rule-builder</h1>
  <div class="liferay-ddm-form-rule-rules-list-container">
    <div class="ddm-rule-list-container form-builder-rule-list">
      <ul class="ddm-form-body-content form-builder-rule-builder-rules-list tabular-list-group">
        <li class="list-group-item">
          <div class="clamp-horizontal list-group-item-content">
            <p class="form-builder-rule-builder-rule-description text-default"><b>if </b>
              <div class="ddm-condition-container"><span>field </span><span class="label label-lg label-secondary" data-original-title="label text 1" title="label text 1"><span class="text-truncate-inline"><span class="text-truncate">label text 1</span></span></span></div><b class="text-lowercase"><em> contains </em></b>
              <div class="ddm-condition-container"><span class="label label-lg label-secondary" data-original-title="" title=""><span class="text-truncate-inline"><span class="text-truncate"></span></span></span></div><br><b> or </b>
              <div class="ddm-condition-container"><span>field </span><span class="label label-lg label-secondary" data-original-title="label text 1" title="label text 1"><span class="text-truncate-inline"><span class="text-truncate">label text 1</span></span></span></div><b class="text-lowercase"><em> is-equal-to </em></b>
              <div class="ddm-condition-container"><span>field </span><span class="label label-lg label-secondary" data-original-title="label text 2" title="label text 2"><span class="text-truncate-inline"><span class="text-truncate">label text 2</span></span></span></div><br><span><b>require </b></span><span class="label label-lg label-secondary" data-original-title="label text 1" title="label text 1"><span class="text-truncate-inline"><span class="text-truncate">label text 1</span></span></span>
            </p>
          </div>
          <div class="list-group-item-field">
            <div class="card-col-field">
              <div class="invalid-rule" title="due-to-missing-fields"><span class="label label-danger"><span class="label-item label-item-expand">broken-rule</span></span></div>
              <div class="dropdown dropdown-action">
                <div class="ddm-rule-list-settings" data-card-id="0">
                  <div class="dropdown dropdown-action"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" title="actions" type="button"><svg class="lexicon-icon lexicon-icon-ellipsis-v" focusable="false" role="presentation">
                        <use xlink:href="icons.svg#ellipsis-v"></use>
                      </svg></button></div><button class="component-action cursor-pointer dropdown-toggle" data-rule-index="0"><svg class="lexicon-icon lexicon-icon-ellipsis-v" focusable="false" role="presentation">
                      <use xlink:href="icons.svg#ellipsis-v"></use>
                    </svg></button>
                </div>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="edit" type="button">edit</button></li>
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="delete" type="button">delete</button></li>
  </ul>
</div>
`;
